Sources/NIOCertificateHotReloading/NIOCertificateHotReloading.swift

```diff
@@ -0,0 +1,319 @@
+
+        /// The certificate's ``TimedCertificateReloader/Location``.
+        public var location: Location
+
+        /// The certificate's ``TimedCertificateReloader/Encoding``.
+        public var format: Encoding
+        
+        /// Initialize a new ``TimedCertificateReloader/CertificateDescription``.
+        /// - Parameters:
+        ///   - location: A ``TimedCertificateReloader/Location``.
+        ///   - format: A ``TimedCertificateReloader/Encoding``.
+        public init(location: Location, format: Encoding) {
+            self.location = location
+            self.format = format
+        }
+    }
+
+    /// A description of a private key, in terms of its ``TimedCertificateReloader/Location`` and
+    /// ``TimedCertificateReloader/Encoding``.
+    public struct PrivateKeyDescription: Sendable {
+
+        /// The key's ``TimedCertificateReloader/Location``.
+        public var location: Location
+
+        /// The key's ``TimedCertificateReloader/Encoding``.
+        public var format: Encoding
+
+        /// Initialize a new ``TimedCertificateReloader/PrivateKeyDescription``.
+        /// - Parameters:
+        ///   - location: A ``TimedCertificateReloader/Location``.
+        ///   - format: A ``TimedCertificateReloader/Encoding``.
+        public init(location: Location, format: Encoding) {
+            self.location = location
+            self.format = format
+        }
+    }
+
+    private struct CertificateKeyPair {
+        var certificate: NIOSSLCertificateSource
+        var privateKey: NIOSSLPrivateKeySource
+    }
+
+    private let refreshInterval: TimeAmount
+    private let certificateDescription: CertificateDescription
+    private let privateKeyDescription: PrivateKeyDescription
+    private let state: NIOLockedValueBox<CertificateKeyPair?>
+
+    /// A `NIOSSLContextConfigurationOverride` that will be used as part of the NIO application's TLS configuration.
+    /// Its certificate and private key will be kept up-to-date via the reload mechanism the ``TimedCertificateReloader``
+    /// implementation provides.
+    public var sslContextConfigurationOverride: NIOSSLContextConfigurationOverride {
+        get {
+            var override = NIOSSLContextConfigurationOverride()
+            guard let certificateKeyPair = self.state.withLockedValue({ $0 }) else {
+                return override
+            }
+            override.certificateChain = [certificateKeyPair.certificate]
+            override.privateKey = certificateKeyPair.privateKey
+            return override
+        }
+    }
+
+    /// Initialize a new ``TimedCertificateReloader``.
+    /// - Parameters:
+    ///   - refreshInterval: The interval at which attempts to update the certificate and private key should be made.
+    ///   - certificateDescription: A ``CertificateDescription``.
+    ///   - privateKeyDescription: A ``PrivateKeyDescription``.
+    public init(
+        refreshingEvery refreshInterval: TimeAmount,
```

Comment:
nit: from the [API design guidelines](https://www.swift.org/documentation/api-design-guidelines/#strive-for-fluent-usage):

> The first argument to initializer and [factory methods](https://en.wikipedia.org/wiki/Factory_method_pattern) calls should not form a phrase starting with the base name